### PR TITLE
use SELinux to restrict datastore modifications

### PR DESCRIPTION
**Pull request by bcressey**
**Issue number:**
#764 

**Description of changes:**
Adds filesystem labels to the four components that we expect to modify the datastore. Extends the policy with rules to allow the transition to `api_t` and mutations for `private_t` objects. 

Changes `updog` to store its cached repo metadata outside the reserved filesystem.

**Testing done:**
Tested `aws-dev` in a local VM and `aws-k8s-1.15` in an EC2 instance. No SELinux denials observed.

Components had the expected label:
```
# find /usr/bin -context 'system_u:object_r:api_exec_t:s0'
/usr/bin/apiserver
/usr/bin/storewolf
/usr/bin/migrator
/usr/bin/early-boot-config
```

`apiserver` ran with the expected label:
```
$ cat /proc/$(pgrep apiserver)/attr/current
system_u:system_r:api_t:s0
```

`updog` worked and created the files in the expected path:
```
# updog check-update -a
aws-k8s-1.15 0.3.2
aws-k8s-1.15 0.3.1
aws-k8s-1.15 0.3.0

# ls -latrZ /var/cache/bottlerocket-metadata/
total 24
drwxr-xr-x. 3 root root system_u:object_r:local_t:s0 4096 Apr 22 21:01 ..
drwxr-xr-x. 2 root root system_u:object_r:local_t:s0 4096 Apr 22 21:01 .
-rw-r--r--. 1 root root system_u:object_r:local_t:s0 1279 Apr 22 21:02 timestamp.json
-rw-r--r--. 1 root root system_u:object_r:local_t:s0 1471 Apr 22 21:02 snapshot.json
-rw-r--r--. 1 root root system_u:object_r:local_t:s0 3497 Apr 22 21:02 targets.json
-rw-r--r--. 1 root root system_u:object_r:local_t:s0   32 Apr 22 21:02 latest_known_time.json
```

Policy blocked access from a different label:
```
# echo -n 'system_u:system_r:system_t:s0' > /proc/self/attr/current
# touch /var/lib/bottlerocket/stuff
[ 1090.055119] audit: type=1400 audit(1587590229.927:7): avc:  denied  { write } for  pid=13065 comm="touch" path="/var/lib/bottlerocket/stuff" dev=
"nvme0n1p10" ino=19 scontext=system_u:system_r:system_t:s0 tcontext=system_u:object_r:private_t:s0 tclass=file permissive=1
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: sources/updater/updog/src/main.rs [hunk 1]
```diff
@@ -31,6 +31,7 @@ const TARGET_ARCH: &str = "aarch64";
 
 const TRUSTED_ROOT_PATH: &str = "/usr/share/updog/root.json";
 const MIGRATION_PATH: &str = "/var/lib/bottlerocket-migrations";
+const METADATA_PATH: &str = "/var/cache/bottlerocket-metadata";
```
Comment (tjkirch) [inline sources/updater/updog/src/main.rs:34]: Is the idea that we would protect this path with different SELinux policies?  I don't see any in the other commit.  I know it's more complex because different things could call updog, but we don't want random processes being able to change this metadata or verification of other signed artifacts (like migrations) could be thwarted.

File: packages/selinux-policy/fs.cil [hunk 2]
```diff
@@ -43,6 +43,10 @@
 (filecon "/.*/usr/bin/containerd.*" file runtime_exec)
 (filecon "/.*/usr/bin/docker.*" file runtime_exec)
 (filecon "/.*/usr/sbin/runc" file runtime_exec)
+(filecon "/.*/usr/bin/apiserver" file api_exec)
+(filecon "/.*/usr/bin/early-boot-config" file api_exec)
```
Comment (tjkirch) [inline packages/selinux-policy/fs.cil:47]: Could we give early-boot-config a more specific policy that only allows access to create/modify the marker file, and not anything in the data store?  Or could we move the marker file?

File: packages/selinux-policy/fs.cil [hunk 3]
```diff
@@ -43,6 +43,10 @@
 (filecon "/.*/usr/bin/containerd.*" file runtime_exec)
 (filecon "/.*/usr/bin/docker.*" file runtime_exec)
 (filecon "/.*/usr/sbin/runc" file runtime_exec)
+(filecon "/.*/usr/bin/apiserver" file api_exec)
```
Comment (tjkirch) [inline packages/selinux-policy/fs.cil:46]: Would the "all_s" rule in combination with this `.*` wildcard allow someone to create/mount `/local/usr/bin/apiserver` and have it get `api_exec` and `api_t` and bypass our restrictions?

**Review by tjkirch — COMMENTED** [review]
Would you be willing to do an upgrade (just to the same version) to make sure the full updog cycle is OK?

File: packages/selinux-policy/fs.cil (hunk 2 above)
Comment (bcressey) [reply packages/selinux-policy/fs.cil:47]: Right now the datastore's label is applied as a mount option, and I like that property for simplicity: it saves us from needing to worry about pre-existing SELinux labels on upgrade or downgrade.

Regarding the marker file location - this is the only mutable storage we have that isn't on the general storage volume, and we want to allow that volume to be swapped at runtime (#614). So either we can't store our marker file there or we have to make it unique per host in some way.

In short, both are possible, just not sure the added complexity is worth it.

File: sources/updater/updog/src/main.rs (hunk 1 above)
Comment (bcressey) [reply sources/updater/updog/src/main.rs:34]: That would be an extra layer of defense, but we don't strictly need it since our root of trust is immutable.

I'd rather approach that as a different issue since it involves giving `updog` a label of its own (`update_t`) and allowing processes to transition to that label. That'll be much more satisfying once the update API is in place; otherwise, we also have to let the container process labels use that transition, or the operator won't work correctly.

File: packages/selinux-policy/fs.cil (hunk 3 above)
Comment (bcressey) [reply packages/selinux-policy/fs.cil:46]: These statements are used to generate `file_contexts`, which is a text file used by tools like `setfiles`.

This is one of the reasons that the "relabel" actions are all reserved (though still allowed for `super_t`). However, you don't need `file_contexts` from the host to perform the relabel if you're allowed to relabel; it's just a text file and you can write your own with the patterns you want.

**Review by tjkirch — APPROVED** [review]

**Comment by bcressey** [comment]
> Would you be willing to do an upgrade (just to the same version) to make sure the full updog cycle is OK?

I tested a downgrade from this build to v0.3.2, and confirmed the `updog` worked on both sides.

**Review by jamieand — APPROVED** [review]
LGTM
